Amplitude/Amplitude.m

```diff
@@ -815,22 +862,47 @@ - (void)endSession
  */
 - (void)refreshSessionTime:(NSNumber*) timestamp
 {
-    if (_sessionId < 0) {
+    if (![self inSession]) {
         return;
     }
+    [self setLastEventTime:timestamp];
+}
 
+- (void)setPreviousSessionId:(long long) previousSessionId
+{
     @synchronized (_eventsData) {
-        [_eventsData setValue:timestamp forKey:@"previous_session_time"];
+        [_eventsData setValue:[NSNumber numberWithLongLong:previousSessionId] forKey:@"previous_session_id"];
     }
 }
 
-- (void)turnOffSessionLaterExecute
+- (long long)previousSessionId
 {
-    [self runOnBackgroundQueue:^{
-        if (!_sessionStarted) {
-            _sessionId = -1;
+    @synchronized (_eventsData) {
+        NSNumber* previousSessionId = _eventsData[@"previous_session_id"];
+        if (previousSessionId == nil) {
+            return -1;
         }
-    }];
+        return [previousSessionId longLongValue];
+    }
+}
+
+- (void)setLastEventTime:(NSNumber*) timestamp
+{
+    @synchronized (_eventsData) {
+        [_eventsData setValue:timestamp forKey:@"previous_session_time"];
+    }
+}
+
+- (NSNumber*)lastEventTime
+{
+    @synchronized (_eventsData) {
+        return _eventsData[@"previous_session_time"];
```

Comment:
Do you want to do a nil check here like you did for the previousSessionId (line 882)?